File: backend-spring-boot/src/main/java/no/vicx/backend/user/vm/UserVm.java
package no.vicx.backend.user.vm;

import io.swagger.v3.oas.annotations.media.Schema;
import no.vicx.database.user.VicxUser;

import static io.swagger.v3.oas.annotations.media.Schema.RequiredMode.REQUIRED;

@Schema(description = "A view model representing a user, containing essential user details like " +
        "username, name, email, and profile image status.")
public record UserVm(

        @Schema(
                description = "The username of the user.",
                example = "johndoe",
                requiredMode = REQUIRED)
        String username,

        @Schema(
                description = "The name of the user.",
                example = "John Doe",
                requiredMode = REQUIRED)
        String name,

        @Schema(
                description = "The email address of the user.",
                example = "johndoe@example.com",
                requiredMode = REQUIRED)
        String email,

        @Schema(
                description = "Indicates whether the user has a profile image.",
                example = "true",
                requiredMode = REQUIRED)
        boolean hasImage
) {
    public static UserVm fromVicxUser(VicxUser user) {
        return new UserVm(
                user.getUsername(),
                user.getName(),
                user.getEmail(),
                user.getUserImage() != null
        );
    }
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/user/vm/UserVm.kt
package no.vicx.backend.user.vm

import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.media.Schema.RequiredMode
import no.vicx.database.user.VicxUser

@Schema(
    description = "A view model representing a user, containing essential user details like " +
            "username, name, email, and profile image status."
)
data class UserVm(
    @Schema(
        description = "The username of the user.",
        example = "johndoe",
        requiredMode = RequiredMode.REQUIRED
    )
    val username: String = "",

    @Schema(
        description = "The name of the user.",
        example = "John Doe",
        requiredMode = RequiredMode.REQUIRED
    )
    val name: String = "",

    @Schema(
        description = "The email address of the user.",
        example = "johndoe@example.com",
        requiredMode = RequiredMode.REQUIRED
    )
    val email: String = "",

    @Schema(
        description = "Indicates whether the user has a profile image.",
        example = "true",
        requiredMode = RequiredMode.REQUIRED
    )
    val hasImage: Boolean = false
) {
    companion object {
        fun fromVicxUser(user: VicxUser) = UserVm(
            username = user.username,
            name = user.name,
            email = user.email,
            hasImage = user.userImage != null
        )
    }
}